[New tutorial] Sending Events from DynamoDB using Eventbridge Pipes (#915)
build-on-aws/content: tutorials/ddb-streams-to-eb-pipes/index.md

---
title: "Sending Events from DynamoDB using Eventbridge Pipes"
description: "A no-code integration between Amazon DynamoDB and Eventbridge"
tags:
    - event-driven
    - serverless
    - eventbridge
    - dynamodb
authorGithubAlias: kaizadwadia
authorName: Kaizad Wadia
additionalAuthors:
  - authorName: Yifei Ding
date: 2023-11-14
---

As application architectures grow in complexity, the number of events and data sources that need to integrate also increases. This often leads developers to write custom integration code which is time-consuming, difficult to maintain, and prone to errors. Amazon EventBridge Pipes aim to simplify these complex event-driven architectures by allowing different services to be connected without writing custom integration code.

## Overview

> EventBridge Pipes provide a no-code way to automatically replicate events between EventBridge buses, SQS/SNS topics, and DynamoDB Streams. A common serverless pattern is to detect DynamoDB table updates and publish corresponding events. Traditionally this required custom Lambda functions to stream DynamoDB events. With EventBridge Pipes, you can now implement this pattern in minutes without writing any code.

Reviewer: > A common serverless pattern

Is this document somewhere? If so, where are the links?

Author: It is here: https://aws.amazon.com/blogs/database/dynamodb-streams-use-cases-and-design-patterns/. However, it uses Lambda not Pipes.

> This post describes some common use cases you might encounter, along with their design options and solutions, when migrating data from relational data stores to [Amazon DynamoDB](https://aws.amazon.com/dynamodb/).
We will consider how to manage the following scenarios:
How do you set up a relationship across multiple tables in which, based on the value of an item from one table, you update the item in a second table?
_How do you trigger an event based on a particular item change?_

I changed it to "A common use case for DynamoDB streams..." though to be more specific and hyperlinked it to this post.

> EventBridge Pipes provide a no-code way to automatically replicate events between EventBridge buses, SQS/SNS topics, and DynamoDB Streams. A common serverless pattern is to detect DynamoDB table updates and publish corresponding events. Traditionally this required custom Lambda functions to stream DynamoDB events. With EventBridge Pipes, you can now implement this pattern in minutes without writing any code.

Reviewer: If there are scenarios where using a custom Lambda function still applies, you need to detail here. The point is not only to sell product features, but to share architecture guidance.

Author: Changed it
> However, in cases where the desired target is not in the [list of targets supported by Eventbridge Pipes](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-event-target.html), or more complex operations need to be performed during transformation such as enrichment of the data from other sources, a custom Lambda function is still necessary. That being said, Eventbridge Pipes still allows simple code-free integrations for a range of different services including [but not limited to DynamoDB streams](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-event-source.html).

| 🧩 Prerequisites    | [AWS Account](https://aws.amazon.com/resources/create-account/)|
| 📢 Feedback            | <a href="https://pulse.buildon.aws/survey/DEM0H5VW" target="_blank">Any feedback, issues, or just a</a> 👍 / 👎    |
| ⏰ Last Updated     | 2023-11-14                             |

## Walkthrough

### Step 1: Creating a DynamoDB Table and enabling streams

First, login to the AWS Management Console and navigate to the DynamoDB service. Click on "Create Table" to begin setting up the new table.

For Table name, enter "GameScores". For the Primary key, enter "GameId" for the partition key and "GamerTag" for the sort key. These will be used to uniquely identify each item in the table.

Under Table settings, we can customize read/write capacity as needed for our expected workload. The defaults are fine to start.

The default settings are left for the key attributes GameId and GamerTag.

![DynamoDB Table Creation Image]( images/createtable.png "Create the Table")

Once all fields are entered, scroll down and click "Create" to finish creating the GameScores table. The table will be ready to store game score data with the specified schema.

> You can now start inserting items into the table through the console, SDKs, or other AWS services. Make sure the items contain the configured partition key and sort key. The other attributes like Score, Level, and Timestamp will be optional.

Reviewer: Is it possible to provide the steps of doing this in the tutorial? Maybe using the console?

Author: Done,

> However, in cases where the desired target is not in the [list of targets supported by Eventbridge Pipes](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-event-target.html), or more complex operations need to be performed during transformation such as enrichment of the data from other sources, a custom Lambda function is still necessary. That being said, Eventbridge Pipes still allows simple code-free integrations for a range of different services including [but not limited to DynamoDB streams](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-event-source.html).

> To enable streams on the table, in the DynamoDB console select the "GameScores" table and go to the "Exports and Streams" tab. Scroll all the way down to "DynamoDB stream details" and click "Turn on". In the view, select "New and old images" for Stream view type and check Stream enabled. Choose the shard count based on expected workload. Scroll down and click "Enable" to activate streams. The stream ARN can now be used by applications to process changes.

Reviewer: Let the reader know if there are any consequences of doing this. Is there any charges associated? Will it slow it down the table? Detail those things will increase the trust from the reader on your content.

Author: > Keep in mind that enabling streams does not consume any read or write capacity units (RCUs or WCUs), and so will not affect the performance of the table. However, the data in the streams is stored in shards, and this storage is something that is charged seperately from the table charges. You also pay for accessing the stream records. If you enable streams on an on-demand capacity mode table, you pay per request for the stream records. If the table uses provisioned capacity, stream requests are free but you pay for shard storage and data transfer. Stream records are kept for 24 hours by default. So shards will accumulate over time as records expire. Make sure to delete the table when no longer needed to stop incurring shard storage charges.

First, open the Amazon EventBridge console in AWS. In the left navigation pane, click "Event Buses". On this page, click the "Create event bus" button. Custom event buses allow for greater isolation, access control, organization, separation of concerns, event retention control, and insulation from changes compared to using the shared default event bus. This is because the default event bus receives events from all AWS services.

Give the event bus the name "game-bus". The name can contain up to 256 characters and must be unique within the account. You can enable archive events here if desired. Click "Create".

Now we will create a rule that sends all events from that bus into a CloudWatch log group so that we can see the events published to it. Go to "Rules" on the left hand side and once there, change the event bus from the default one to "game-bus".

![Eventbridge Rules]( images/rules.png "Eventbridge Rules")

Click on "Create Rule" and for a rule name, enter "all-game-events". Optionally enter a description. Make sure the event bus selected is "game-bus".

![Eventbridge Rule Step 1]( images/createrules1.png "Eventbridge Rule Step 1")

> After going to the next step, click on "All Events" as the event source. This may display a warning but we do want to have visibility into whatever events are sent to the bus. We can leave everything else in this step as is and continue.

Reviewer: > This may display a warning

Which warning? Please detail in the tutorial what the reader should expect to see.

Author: The warning is shown in the image. Updated to make it clearer

> 
After going to the next step, click on "All Events" as the event source. This may display a warning (as displayed in the image below) but we do want to have visibility into whatever events are sent to the bus. The warning is shown because rules are generally used to filter events coming into the Event bus, and our rule is doing no filtering. You may create rules that performs filtering but for this example we will perform filtering prior to adding the event to the event bus and  We can leave everything else in this step as is and continue.

![Eventbridge Rule Step 2]( images/createrules1.png "Eventbridge Rule Step 2")

> For the target, select "AWS service" and for the target, from the dropdown select "CloudWatch log group". Name the log group "game-events-log". Note that if we use infrastructure as code, we would additionally need to configure an IAM role to allow Eventbridge to access this log group.

Reviewer: Tutorial must keep the consistency on what approach they are leading the reader. If everything is being done via the console, you should abstain from mentioning a different approach. This creates more confusion than benefit.

Author: Removed that part

![Eventbridge Rule Step 3]( images/createrules3.png "Eventbridge Rule Step 3")

Now we can skip the tagging part and go straight to creating the rule.

![Eventbridge Rule Description]( images/ruledescription.png "Eventbridge Rule Description")

Once the rule is created, we can navigate to it in the console under the "Rules" section. Then, if we navigate to the "targets" tab on the lower side of the page, we can then click the link that says "game-events-log" to navigate to the created log group in CloudWatch. Here, we can see a log of all the events sent to the event bus. We can click on "Start tailing" on the top right hand side to see the incoming events. It is highly recommended to leave this tab open as we will return later.

> ### Step 3: Creating the pipe

Reviewer: I think this section could use a more detail explanation about the concept of pipe. Maybe something as simple as explaining that the pipe is responsible for the actions that will carry over the events through the buses and apply data mutation operations such as transformations, filtering, etc. Always assume the reader doesn't know everything about the product.

Author: > We will now create the Eventbridge Pipe. Eventbridge Pipes are a simple, no-code integration service that connects sources to targets. Pipes also supports more advanced capabilities including transformations as well as enrichment using an API, a Step Function or an AWS Lambda function. We will use Pipes to read data from the DynamoDB stream and write it to the custom event bus we created in the last section.

To perform filtering, we may go back to the "game-event-pipe" in the Eventbridge Pipes section of the Eventbridge console. Then, we can click on "Edit" on the top right hand side to edit the pipe. Then clicking on the "Filtering" icon on the pipe should get us the ability to add a filtering statement. Copy the following event pattern into the box that says "Event Pattern" near the bottom of the page.

```json
{
  "dynamodb": {
    "Keys": {
      "GameId": {
        "S": [ { "equals-ignore-case": "1757" } ]
      }
    }
  }
}
```

![Filtering Events]( images/filtering.png "Filtering Events")

Now that we configured filtering, let's transform the payload using the "Target Input Transformer". To do this, click on the "Target" Icon on the pipe diagram and scroll down. Close to the bottom there is a dropdown called "Target Input Transformer", which should be expanded. Paste the following transformer in the middle column:

```

> {

Reviewer: It would be beneficial for the reader to explain the syntax of this JSON document. Moreover, provides links that would help the reader to learn more about advanced examples.

We can now play around with the Table to see how the events get recorded. Navigate to the DynamoDB table and change the fields in the old item we placed in the table, like how we did in the previous section. Notice how it got sent because the filtering statement matched the GameId. How about if we create another item with another GameId? We see that gets filtered out. Now any new event that gets sent through Pipes will now get transformed to a format that looks similar to this one, instead of one that looks like the default output of a DynamoDB stream:

```json
{
    "version": "0",
    "id": "8b5b7650-5f62-128b-4ca4-18e06ff9e4bb",
    "detail-type": "Event from aws:dynamodb",
    "source": "Pipe game-event-pipe",
    "account": "AWS_ACCOUNT_ID",
    "time": "2023-11-12T15:01:02Z",
    "region": "eu-west-1",
    "resources": [],
    "detail": {
        "pipeName": "game-event-pipe",
        "GamerTag": "johndoe123",
        "GameId": "1757",
        "OldScore": "42122",
        "NewScore": "42123",
        "OldLevel": "8",
        "NewLevel": "9"
    }
}
```

> We have now learned how to transform and filter the inputs. Notice that there is also the Enrichment feature. This is useful if you want to enrich the data using a Lambda function, a Step Function or another API that can perform more complex transformations or even add data from other sources before it gets returned to the pipe.

Reviewer: This is a great place to lead the reader to consume another content that provides the steps to learn about event enrichment.

Author: > We will now create the Eventbridge Pipe. Eventbridge Pipes are a simple, no-code integration service that connects sources to targets. Pipes also supports more advanced capabilities including transformations as well as enrichment using an API, a Step Function or an AWS Lambda function. We will use Pipes to read data from the DynamoDB stream and write it to the custom event bus we created in the last section.

Author: Removing this part because we don't have another data source so it would essentially just be a transformation. I also don't want to have any custom code here to keep it simple (hence this section also being optional, even though the "transformer" isn't really code)

## Clean-up

To clean-up your AWS account, remember to delete the following:

* The CloudWatch log group in the CloudWatch Console
* The Eventbridge pipe, as well as the custom bus and the rule.
* The DynamoDB Table. This will automatically delete the stream. Make sure to not check the box for "Create an on-demand backup of GameScores before deletion."

## Conclusion

> Through this step-by-step guide, we created an end-to-end serverless event pipeline using Amazon EventBridge Pipes. By simply pointing and clicking in the AWS Management Console, we set up a Pipe that streams DynamoDB table update events to an EventBridge event bus. This removes the need to write any custom integration logic in Lambda, allowing us to focus on application functionality instead of plumbing.

Reviewer: > instead of plumbing

I would be careful with the word plumbing. In practice, this is still a lot of plumbing to get things done, just not one that is done programmatically. I would just say: "focus on the application functionality."

Author: True, removed this

> Pipes simplified several key tasks for our use case: capturing DynamoDB events, filtering and transforming the event payload, and routing events to the target EventBridge bus. We also saw how to optionally enrich events by invoking a Lambda function.

Reviewer: Did it? I saw the example of filtering and transformation, but not the one about enrichment. You have mentioned is possible, but not necessarily provided an example.

Author: Removed

>Pipes simplified several key tasks for our use case: capturing DynamoDB events, routing events to the target EventBridge bus, and optionally filtering and transforming the event payload.

> ---

Reviewer 2: Hello! 

Nice opening paragraph of stating the problem, the consequence of the problem and the solution. Well done! The structure, narrative pacing and tone are solid. 

I corrected titles to AP Style and proper names. Please review that all proper names begin with a capital letter. Most notably, “Amazon EventBridge” is sometimes “Amazon Eventbridge”. The former is the correct legal brand name. Please rake through the piece and make the corrections. 

Please see my notes below for more information.

> title: "Sending Events from DynamoDB using Eventbridge Pipes"

Reviewer 2: ```suggestion
title: "Sending Events From DynamoDB Using EventBridge Pipes"
```

> description: "A no-code integration between Amazon DynamoDB and Eventbridge"

Reviewer 2: Good job making this descriptor fewer than 50 characters. But, it can't repeat the messaging on the title or use the same phrases, words. It should also focus on what solution the reader gets from reading the piece. I wrote an example below but feel free to rewrite. 

```suggestion
description: "Save time with a manageable no-code integration to replicate events."

```

> tags:

Reviewer 2: We need 5 tags. Please add 1 more.

![DynamoDB Streams Enable]( images/enablestreams.png "Enabling Streams")

### Step 2: Create a custom event bus in Amazon Eventbridge

First, open the Amazon EventBridge console in AWS. In the left navigation pane, click "Event Buses". On this page, click the "Create event bus" button. Custom event buses allow for greater isolation, access control, organization, separation of concerns, event retention control, and insulation from changes compared to using the shared default event bus. This is because the default event bus receives events from all AWS services.

Give the event bus the name "game-bus". The name can contain up to 256 characters and must be unique within the account. You can enable archive events here if desired. Click "Create".

Now we will create a rule that sends all events from that bus into a CloudWatch log group so that we can see the events published to it. Go to "Rules" on the left hand side and once there, change the event bus from the default one to "game-bus".

> ![Eventbridge Rules]( images/rules.png "Eventbridge Rules")

Reviewer 2: Please make sure it's "EventBridge" as it's a brand name. 

```suggestion
![EventBridge Rules]( images/rules.png "EventBridge Rules")
```

Click on "Create Rule" and for a rule name, enter "all-game-events". Optionally enter a description. Make sure the event bus selected is "game-bus".

![Eventbridge Rule Step 1]( images/createrules1.png "Eventbridge Rule Step 1")

> After going to the next step, click on "All Events" as the event source. This may display a warning (as displayed in the image below) but we do want to have visibility into whatever events are sent to the bus. The warning is shown because rules are generally used to filter events coming into the Event bus, and our rule is doing no filtering. You may create rules that performs filtering but for this example we will perform filtering prior to adding the event to the event bus. This rule is just to have visibility into the event bus. We can leave everything else in this step as is and continue.

Reviewer 2: ```suggestion
After going to the next step, click on "All Events" as the event source. This may display a warning (as displayed in the image below) but we do want to have visibility into whatever events are sent to the bus. The warning is shown because rules are generally used to filter events coming into the Event bus, and our rule is doing no filtering. You may create rules that performs filtering but for this example we will perform filtering prior to adding the event to the Event bus. This rule is just to have visibility into the Event bus. We can leave everything else in this step as is and continue.
```

![Eventbridge Rule Step 2]( images/createrules1.png "Eventbridge Rule Step 2")

For the target, select "AWS service" and for the target, from the dropdown select "CloudWatch log group". Name the log group "game-events-log".

![Eventbridge Rule Step 3]( images/createrules3.png "Eventbridge Rule Step 3")

Now we can skip the tagging part and go straight to creating the rule.

![Eventbridge Rule Description]( images/ruledescription.png "Eventbridge Rule Description")

Once the rule is created, we can navigate to it in the console under the "Rules" section. Then, if we navigate to the "targets" tab on the lower side of the page, we can then click the link that says "game-events-log" to navigate to the created log group in CloudWatch. Here, we can see a log of all the events sent to the event bus. We can click on "Start tailing" on the top right hand side to see the incoming events. It is highly recommended to leave this tab open as we will return later.

> ### Step 3: Creating the pipe

Reviewer 2: ```suggestion
### Step 3: Creating the Pipe
```

> We will now create the Eventbridge Pipe. Eventbridge Pipes are a simple, no-code integration service that connects sources to targets. Pipes also supports more advanced capabilities including transformations as well as enrichments using an API, a Step Function or a Lambda function. We will use Pipes to read data from the DynamoDB stream and write it to the custom event bus we created in the last section.

Reviewer 2: ```suggestion
We will now create the Eventbridge Pipe. Eventbridge Pipes are a simple, no-code integration service that connects sources to targets. Pipes also supports more advanced capabilities including transformations as well as enrichments using an API, a Step function or a Lambda function. We will use Pipes to read data from the DynamoDB stream and write it to the custom event bus we created in the last section.
```

To create the Eventbridge Pipe, we can click on "Pipes" on the left hand side of the Eventbridge console, and click on "Create Pipe". In the name field, we can use "game-event-pipe". For the source, we want to use the DynamoDB stream we previously created. The starting position can be left as "Latest". We could also optionally configure additional settings pertaining to batching if we want to process multiple items at once, but for now we will leave these settings as they are.

![Eventbridge Pipe Source]( images/ruledescription.png "Eventbridge Pipe Source")

Now for the target, we can select our "game-bus" event bus. Now we are ready and can click on "Create Pipe". Now we will test the pipe by creating an item in the DynamoDB Table.

![Eventbridge Pipe Target]( images/pipetarget.png "Eventbridge Pipe Target")

> ### Step 4: Testing the integration

Reviewer 2: ```suggestion
### Step 4: Testing the Integration
```

To test the integration, we can once again navigate to our table in the DynamoDB console. We can do that by going back to the "Tables" page on the console and click on our table called "GameScores". Then we can click on "Explore Table Items" in the top right hand side and after scrolling down click on "Create Item". Then create the item, using the "JSON view" and disabling "View DynamoDB JSON". We can create this item as a sample:

```json
{
  "GameId": "1757",
  "GamerTag": "johndoe123", 
  "Score": 42120,
  "Level": 8,
  "Timestamp": 1613584792
}
```

![Create DynamoDB Item]( images/createddbitem.png "Create DynamoDB Item")

After creating this item, we can then go to the CloudWatch Logs Live Tail, and see that the item is sent to the CloudWatch log group. This is super powerful, as the item had to traverse the stream, into the Eventbridge Pipe, through the event bus, and passed through the Eventbridge Rule in order to get here!

![Event in log]( images/eventinlog.png "Event in log")

> ### Optional: Filtering and transforming the payload

Reviewer 2: ```suggestion
### Optional: Filtering and Transforming the Payload
```

To perform filtering, we may go back to the "game-event-pipe" in the Eventbridge Pipes section of the Eventbridge console. Then, we can click on "Edit" on the top right hand side to edit the pipe. Then clicking on the "Filtering" icon on the pipe should get us the ability to add a filtering statement. Copy the following event pattern into the box that says "Event Pattern" near the bottom of the page. This event pattern follows the same pattern as the expected event except for the fact that the value is replaced by an array detailing the kind of value that is expected. In this case, we case-insensitively expect the string "1757" for the GameId string.

```json
{
  "dynamodb": {
    "Keys": {
      "GameId": {
        "S": [ { "equals-ignore-case": "1757" } ]
      }
    }
  }
}
```

![Filtering Events]( images/filtering.png "Filtering Events")

> Now that we configured filtering, let's transform the payload using the "Target Input Transformer". To do this, click on the "Target" Icon on the pipe diagram and scroll down. Close to the bottom there is a dropdown called "Target Input Transformer", which should be expanded. Paste the following transformer in the middle column:

Reviewer 2: ```suggestion
Now that we configured filtering, let's transform the payload using the "Target Input Transformer". To do this, click on the "Target" icon on the pipe diagram and scroll down. Close to the bottom there is a dropdown called "Target Input Transformer", which should be expanded. Paste the following transformer in the middle column:
```

```
{
    "pipeName" : <aws.pipes.pipe-name>,
    "GamerTag": <$.dynamodb.Keys.GamerTag.S>,
    "GameId": <$.dynamodb.Keys.GameId.S>,
    "OldScore": <$.dynamodb.OldImage.Score.N>,
    "NewScore": <$.dynamodb.NewImage.Score.N>,
    "OldLevel": <$.dynamodb.OldImage.Level.N>,
    "NewLevel": <$.dynamodb.NewImage.Level.N>
}
```

An [input transformer](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-input-transformation.html) object looks syntatically similar to a JSON document, except for the fact that the dynamic values inside the "<>" tags are replaced by the values from the event and metadata from the pipe, such as the pipe name. Hence, these do not use quotes. Static values can also be inserted with the usual JSON syntax.

![Transforming Events]( images/transformer.png "Transforming Events")

If you like, you can try pasting a sample event from CloudWatch Logs on the right hand side column to see how the event would get transformed. This should take information from the item, as well as metadata from the pipe and transform the event before it is sent to the event bus. Now we can click "Update pipe" on the top right hand side.

> We can now play around with the Table to see how the events get recorded. Navigate to the DynamoDB table and change the fields in the old item we placed in the table, like how we did in the previous section. Notice how it got sent because the filtering statement matched the GameId. How about if we create another item with another GameId? We see that gets filtered out. Now any new event that gets sent through Pipes will now get transformed to a format that looks similar to this one, instead of one that looks like the default output of a DynamoDB stream:

Reviewer 2: ```suggestion
We can now play around with the table to see how the events get recorded. Navigate to the DynamoDB table and change the fields in the old item we placed in the table, like how we did in the previous section. Notice how it got sent because the filtering statement matched the GameId. How about if we create another item with another GameId? We see that gets filtered out. Now any new event that gets sent through Pipes will now get transformed to a format that looks similar to this one, instead of one that looks like the default output of a DynamoDB stream:
```

```json
{
    "version": "0",
    "id": "8b5b7650-5f62-128b-4ca4-18e06ff9e4bb",
    "detail-type": "Event from aws:dynamodb",
    "source": "Pipe game-event-pipe",
    "account": "AWS_ACCOUNT_ID",
    "time": "2023-11-12T15:01:02Z",
    "region": "eu-west-1",
    "resources": [],
    "detail": {
        "pipeName": "game-event-pipe",
        "GamerTag": "johndoe123",
        "GameId": "1757",
        "OldScore": "42122",
        "NewScore": "42123",
        "OldLevel": "8",
        "NewLevel": "9"
    }
}
```

We have now learned how to transform and filter the inputs.

> ## Clean-up

Reviewer 2: ```suggestion
## Clean Up
```

> To clean-up your AWS account, remember to delete the following:

Reviewer 2: ```suggestion
To clean up your AWS account, remember to delete the following:
```

* The CloudWatch log group in the CloudWatch Console
* The Eventbridge pipe, as well as the custom bus and the rule.
* The DynamoDB Table. This will automatically delete the stream. Make sure to not check the box for "Create an on-demand backup of GameScores before deletion."

## Conclusion

Through this step-by-step guide, we created an end-to-end serverless event pipeline using Amazon EventBridge Pipes. By simply pointing and clicking in the AWS Management Console, we set up a Pipe that streams DynamoDB table update events to an EventBridge event bus. This removes the need to write any custom integration logic in Lambda, allowing us to focus on application functionality.

Pipes simplified several key tasks for our use case: capturing DynamoDB events, routing events to the target EventBridge bus, and optionally filtering and transforming the event payload.

> By leveraging EventBridge Pipes for event ingestion and integration, we can quickly connect various services and data sources without managing complex application code. Pipes provide a no-code way to implement event streaming and transformation in our serverless architectures. Going forward, we can spend less time on glue code and more time focusing on core product capabilities.

Reviewer 2: Beautiful conclusion! But, we want to invite readers to stay in the blog repository longer. Please add a sentence or two that is apropos of your topic and with 2-3 blog links.  

EX: "To learn more about X, visit (LINK) and (LINK)."
